calico_node/startup.go

```diff
@@ -14,59 +14,657 @@
+	pool := &api.IPPool{
+		Metadata: api.IPPoolMetadata{
+			CIDR: *cidr,
+		},
+		Spec: api.IPPoolSpec{
+			NATOutgoing: true,
+		},
+	}
+	if _, err := client.IPPools().Apply(pool); err != nil {
+		if _, ok := err.(errors.ErrorResourceAlreadyExists); !ok {
+			fatal("Failed to create default IPv%d IP pool: %s", version, err)
+			terminate()
+		}
+	}
+	message("Created default IPv%d pool (%s) with NAT outgoing enabled", version, cidr)
+}
+
+// Check whether any other nodes have been configured with the same IP addresses.
+func checkConflictingNodes(client *client.Client, node *api.Node) {
+	// Get the full set of nodes.
+	var nodes []api.Node
+	if nodeList, err := client.Nodes().List(api.NodeMetadata{}); err != nil {
+		fatal("Unable to query node confguration: %s", err)
+		terminate()
+	} else {
+		nodes = nodeList.Items
+	}
+
+	ourIPv4 := node.Spec.BGP.IPv4Address
+	ourIPv6 := node.Spec.BGP.IPv6Address
+	errored := false
+	for _, theirNode := range nodes {
+		if theirNode.Spec.BGP == nil {
+			// Skip nodes that don't have BGP configured.  We know
+			// that this node does have BGP since we only perform
+			// this check after configuring BGP.
+			continue
+		}
+		theirIPv4 := theirNode.Spec.BGP.IPv4Address
+		theirIPv6 := theirNode.Spec.BGP.IPv6Address
+
+		// If this is our node (based on the name), check if the IP
+		// addresses have changed.  If so warn the user as it could be
+		// an indication of multiple nodes using the same name.  This
+		// is not an error condition as the IPs could actually change.
+		if theirNode.Metadata.Name == node.Metadata.Name {
+			if theirIPv4 != nil && !theirIPv4.Equal(ourIPv4.IP) {
+				warning("Calico node '%s' IPv4 address has changed:",
+					theirNode.Metadata.Name)
+				message(" -  This could happen if multiple nodes are configured with the same name")
+				message(" -  Original IP: %s", theirIPv4)
+				message(" -  Updated IP: %s", ourIPv4)
+			}
+			if theirIPv6 != nil && ourIPv6 != nil && !theirIPv6.Equal(ourIPv6.IP) {
+				warning("Calico node '%s' IPv6 address has changed:",
+					theirNode.Metadata.Name)
+				message(" -  This could happen if multiple nodes are configured with the same name")
+				message(" -  Original IP: %s", theirIPv6)
+				message(" -  Updated IP: %s", ourIPv6)
+			}
+			continue
+		}
+
+		// Check that other nodes aren't using the same IP addresses.
+		// This is an error condition.
+		if theirIPv4 != nil && theirIPv4.Equal(ourIPv4.IP) {
+			message("Calico node '%s' is already using the IPv4 address %s:",
+				theirNode.Metadata.Name, ourIPv4)
+			message(" -  Check the node configuration to remove the IP address conflict")
+			errored = true
+		}
+		if theirIPv6 != nil && theirIPv6.Equal(ourIPv6.IP) {
+			message("Calico node '%s' is already using the IPv6 address %s:",
+				theirNode.Metadata.Name, ourIPv6)
+			message(" -  Check the node configuration to remove the IP address conflict")
+			errored = true
+		}
+	}
+
+	if errored {
```

Comment:
We can, but it means that we won't handle IPv4 and IPv6 conflicts if the conflicts are on different nodes - so we'll only log the first error and then exit.  I think how it is is probably the "correct" way.